infrastructure/src/main/java/org/corfudb/infrastructure/CompactorLeaderServices.java

```diff
@@ -0,0 +1,334 @@
+                        "null" : managerStatus.getStatus());
+                txn.commit();
+                return;
+            }
+
+            List<TableName> tableNames = new ArrayList<>(txn.keySet(compactorMetadataTables.getCheckpointingStatusTable()));
+            StatusType finalStatus = StatusType.COMPLETED;
+            for (TableName table : tableNames) {
+                CheckpointingStatus tableStatus = (CheckpointingStatus) txn.getRecord(
+                        CompactorMetadataTables.CHECKPOINT_STATUS_TABLE_NAME, table).getPayload();
+                StringBuilder str = new StringBuilder();
+                str.append(printCheckpointStatus(table, tableStatus));
+                syslog.info("{}", str);
+                if (tableStatus.getStatus() != StatusType.COMPLETED) {
+                    finalStatus = StatusType.FAILED;
+                    break;
+                }
+            }
+            long totalTimeElapsed = System.currentTimeMillis() - managerStatus.getTimeTaken();
+            txn.putRecord(compactorMetadataTables.getCompactionManagerTable(), CompactorMetadataTables.COMPACTION_MANAGER_KEY,
+                    buildCheckpointStatus(finalStatus, tableNames.size(), totalTimeElapsed, managerStatus.getEpoch()),
+                    null);
+            txn.commit();
+            syslog.info("Total time taken for the compaction cycle: {}ms for {} tables with status {}", totalTimeElapsed,
+                    tableNames.size(), finalStatus);
+            MicroMeterUtils.time(Duration.ofMillis(totalTimeElapsed), "compaction.total.timer",
+                    "nodeEndpoint", nodeEndpoint);
+        } catch (RuntimeException re) {
+            //Do not retry here, the compactor service will trigger this method again
+            syslog.warn("Exception in finishCompactionCycle: {}. StackTrace={}", re, re.getStackTrace());
+        }
+        deleteInstantKeyIfRequired();
+        trimLogIfRequired();
+    }
+
+    private void trimLogIfRequired() {
+        Optional<RpcCommon.TokenMsg> upgradeToken = Optional.empty();
+        for (int i = 0; i < MAX_RETRIES; i++) {
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+                upgradeToken = Optional.ofNullable((RpcCommon.TokenMsg) txn.getRecord(CompactorMetadataTables.CHECKPOINT,
+                        CompactorMetadataTables.UPGRADE_KEY).getPayload());
+                txn.delete(CompactorMetadataTables.CHECKPOINT, CompactorMetadataTables.UPGRADE_KEY);
+                txn.commit();
+                break;
+            } catch (RuntimeException re) {
+                if (DistributedCheckpointer.isCriticalRuntimeException(re, i, MAX_RETRIES)) {
+                    break;
+                }
+            }
+        }
+        if (upgradeToken.isPresent()) {
+            syslog.info("Upgrade Key found: Hence invoking trimlog()");
+            trimLog.invokePrefixTrim();
```

Comment:
Since no corfu client (other services) is running during upgrade, it's safe to trim log right after checkpointing. This allows to run compactor only once during data migration, which reduces the overall upgrade time - I think upgrade team will be happy to see this!